src/lib.rs

```diff
@@ -39,14 +39,77 @@
 #![crate_type = "rlib"]
 #![feature(core_intrinsics)]
 #![feature(allocator_api)]
+#![feature(alloc_layout_extra)]
 
 mod alloc;
 
 use crate::alloc::{AllocWithInfo, GCMalloc};
+use std::{
+    alloc::{Alloc, Layout},
+    mem::{forget, size_of},
+    ptr,
+};
 
 #[global_allocator]
 static ALLOCATOR: AllocWithInfo = AllocWithInfo;
 
 /// Used for allocation of objects which are managed by the collector (through
 /// the `Gc` smart pointer interface).
 static mut GC_ALLOCATOR: GCMalloc = GCMalloc;
+
+struct Gc<T> {
+    objptr: *mut T,
+}
+
+impl<T> Gc<T> {
+    pub fn new(v: T) -> Self {
+        let objptr = Self::alloc_blank(Layout::new::<T>());
+        let gc = unsafe {
+            objptr.copy_from_nonoverlapping(&v, 1);
+            Gc::from_raw(objptr)
+        };
+        forget(v);
+        gc
+    }
+
+    /// Create a `Gc` from a raw pointer previously created by `alloc_blank` or
+    /// `into_raw`.
+    pub unsafe fn from_raw(objptr: *const T) -> Self {
+        Gc {
+            objptr: objptr as *mut T,
+        }
+    }
+
+    /// Allocate memory sufficient to `l` (i.e. correctly aligned and of at
+    /// least the required size). The returned pointer must be passed to
+    /// `Gc::from_raw`.
+    pub fn alloc_blank(l: Layout) -> *mut T {
+        let (layout, uoff) = Layout::new::<usize>().extend(l).unwrap();
+
+        unsafe {
+            let baseptr = GC_ALLOCATOR.alloc(layout).unwrap().as_ptr();
+            let objptr = baseptr.add(uoff);
+            let headerptr = objptr.sub(size_of::<usize>());
```

Comment:
It might be worth copying over (with modifications if necessary) the comment from `abgc` that explains why we do this particular dance?

Reply:
Done